### DISCUSSION TITLE
Items to close down phase 0 (PoC)

### DISCUSSION THREAD

**Discussion by adesutherland** [adesutherland/CREXX#123] [Progress Updates]
Very short term: I have taken card "Feature f0023 - show hex of instructions in disassembler" because it is a nice quick task

Medium term: This is what I think we need to do to take us to the end of the PoC phase 0, in priority order. This is just to put a stake in the ground - the whole team needs to agree the list and order, but I think it would be useful to have a list. The point is that it is quite a long list but actually not too long ... I think we can reach the end! 

* Optional Procedure Arguments - this will allow BiFs to be completed / tested

* Classic REXX BiFs with Automatic Tests written in REXX level B with new RXAS Instructions as needed

* Keyhole optimiser. This will allow Peter's super-instructions to be generated. Moreover the actual replacement rules will be extensible (and quite simple) and I think anyone in the team will be able to add rules and instructions and see the compiled code using them without having to be an expert in the compiler internals.

* Performance Tests / Baselines. Including porting or making the equivalent of the the classic REXX test

* Platforms: Windows, Mac, Linux, VM/370, MVS/370, and S/390 (I mean modern mainframes so we can compare performance)

* INTERPRET Support. This is a technical risk which we should tackle early to avoid re-work.

* cREXX Classes. This will give us the semantics (or perhaps I mean syntax) to create the entire REXX level C runtime in REXX including BiFs, Variable Pools and Error Messages. 

* Build tools for libraries. Packaging the above

* Enhance REXX Level B syntax to have the same capabilities as classic REXX (e.g. more loops and select)

* REXX Level C (subset) front-end. So we have classic REXX syntax support

* Documentation. cREXX Principles of Operations, RXAS Programmers Guide, REXX Level B Programmers Guide, cREXX Level C (subset) Programmers Guide  

At this stage, we should have a "just about" classic REXX implementation with a front end compiler, vm and runtime library. 

It wont be complete, for example PARSE and DECIMAL are the major missing's, and we may need to include them in the scope for phase 0, in any case they will probably be the first job in phase 1. Perhaps we need to have "rough and ready" implementations of these to show the approach?

Other tasks post phase 0 include compiler optimisations (flow/data analysis), multiple platforms, alternative backends (LLVM, JavaScript), Level G Runtime library, full ANSI Classic REXX, and so on ...

Does this form a basis for a plan?

**Comment by rvjansen** [comment]
I think that is an excellent basis of a plan. I would have liked to do more today, but it is my wedding anniversary (1992 - 29 years already) - so my attention is needed elsewhere.
